File: 270__Closest_Binary_Search_Tree_Value.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def closestValue(self, root, target):
        """
        :type root: TreeNode
        :type target: float
        :rtype: int
        """
        self.minval = float("inf")
        def helper(root):
            if root is None:
                return
            if abs(root.val - target) < abs(self.minval - target):
                self.minval = root.val
            if root.val > target:
                helper(root.left)
            elif root.val < target:
                helper(root.right)
        helper(root)
        return self.minval

File: Java/270__Closest_Binary_Search_Tree_Value.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int closestValue(TreeNode root, double target) {
        int min_val = root.val, max_val = root.val;
        TreeNode curr = root;
        while (curr != null) {
            if (curr.val < target) {
                min_val = curr.val;
                curr = curr.right;
            }
            else {
                max_val = curr.val;
                curr = curr.left;
            }
        }
        return Math.abs(target - min_val) < Math.abs(max_val - target) ? min_val : max_val;

    }
}
